kube-client/src/api/subresource.rs

```diff
@@ -128,6 +128,178 @@ where
     }
 }
 
+// ----------------------------------------------------------------------------
+// Ephemeral containers
+// ----------------------------------------------------------------------------
+
+/// Marker trait for objects that support the ephemeral containers sub resource.
+pub trait Ephemeral {}
+
+impl Ephemeral for k8s_openapi::api::core::v1::Pod {}
+
+impl<K> Api<K>
+where
+    K: Clone + DeserializeOwned + Ephemeral,
+{
+    /// Replace the ephemeral containers sub resource entirely
+    ///
+    /// This functions in the same way as [`Api::replace`] except only `.spec.ephemeralcontainers` is replaced, everything else is ignored.
+    ///
+    /// Note that ephemeral containers may **not** be changed or removed after being attached to a pod.
+    ///
+    /// See the Kubernetes [documentation](https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/#what-is-an-ephemeral-container)
+    /// for more details.
+    ///
+    /// Example of using `replace_ephemeral_containers`:
+    ///
+    /// ```no_run
+    /// use k8s_openapi::api::core::v1::Pod;
+    /// #[tokio::main]
+    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    ///     let client = Client::try_default().await?;
+    ///     let pods: Api<Pod> = Api::namespaced(client, "apps");
+    ///     let pp = PostParams::default();
+    ///
+    ///     // Get pod object with ephemeral containers.
+    ///     let mut mypod = pods.get_ephemeral_containers("mypod").await?;
+    ///     
+    ///     // If there were existing ephemeral containers, we would have to append
+    ///     // new containers to the list before calling replace_ephemeral_containers.
+    ///     assert_eq(mypod.spec.ephemeral_containers, None);
+    ///
+    ///     // Add an ephemeral container to the pod object.
+    ///     mypod.spec.ephemeral_containers = Some(serde_json::from_value(serde_json::json!([
+    ///        {
+    ///            "name": "myephemeralcontainer",
+    ///            "image": "busybox:1.34.1",
+    ///            "command": ["sh", "-c", "sleep 20"],
+    ///        },
+    ///     ]))?);
+    ///
+    ///     pods.replace_ephemeral_containers("mypod", &pp, serde_json::to_vec(&o)?).await?;
+    ///
+    ///     Ok(())
+    /// }
+    /// ```
+    pub async fn replace_ephemeral_containers(
+        &self,
+        name: &str,
+        pp: &PostParams,
+        data: Vec<u8>,
+    ) -> Result<K> {
```

Comment:
That makes more sense it felt very unergonomic. The other replace_* methods for sub resources do use Vec<u8> though, is the inconsistency alright?

Reply:
ugh, yes, those should also follow the same pattern ideally. i think those are safe to change in a non-breaking way as well though 🤔

Reply:
If we use `K`I think this would become a breaking change in the other methods as `K`is a resource.

This line `serde_json::to_vec(&json!(first_patch))?` would no longer work.

Reply:
Future note: the choice of having all signatures require a `K` still feels sensible to me, but dependent on how portable sending a full K is everywhere. our suspicion here is the apiserver just discards everything not relevant, in which case great. Pod is fully optional so it's not harder for users to pass a K, so we should keep `K` for ephemeral. Coercing to a `K` is a nice way to get type-safety and it means passing less vectors around. BUT whether it's generally better (for all subresources) depends on whether we have [apply-configurations](https://github.com/kube-rs/kube/issues/649#issuecomment-939987209) for all structs. 

If is actually slightly problematic for `CRDs` atm because they hardcode `.spec` as non-optional. It only applies to the general `patch_status` call (which uses a vector rather than a `K`) because `patch_scale` sidesteps the issue entirely by having a separate [`Scale`](https://docs.rs/k8s-openapi/latest/k8s_openapi/api/autoscaling/v1/struct.Scale.html) struct. If we made `patch_status` take a `K`, then we would need to make `spec` optional because otherwise it'd be impossible to create patches for it (without requiring `Default` on spec or something else wacky).

Reply:
That all sounds sensible to me. Just to confirm, are we leaving  `replace_ephemeral_containers` as it is in the current version? meaning it accepts a `K: serializable`

Reply:
Yeah, let's leave this as-is for now. I'll resolve this thread after a follow-up issue.